AsyncDisplayKit/Details/ASCollectionViewLayoutController.mm

```diff
@@ -0,0 +1,153 @@
+  _updateRangeBoundsIndexedByRangeType[rangeType] = rangeGeometry.updateBounds;
+  return [self indexPathsForItemsWithinRangeBounds:rangeGeometry.rangeBounds collectionView:_collectionView];
+}
+
+- (ASRangeGeometry)rangeGeometryWithScrollDirection:(ASScrollDirection)scrollDirection
+                                     collectionView:(ASCollectionView *)collectionView
+                              rangeTuningParameters:(ASRangeTuningParameters)rangeTuningParameters {
+  CGRect rangeBounds = collectionView.bounds;
+  CGRect updateBounds = collectionView.bounds;
+  ASScrollDirection scrollableDirections = [collectionView scrollableDirections];
+  
+  BOOL canScrollHorizontally = ASScrollDirectionContainsHorizontalDirection(scrollableDirections);
+  if (canScrollHorizontally) {
+    ASDirectionalScreenfulBuffer horizontalBuffer = ASDirectionalScreenfulBufferHorizontal(scrollDirection,
+                                                                                           rangeTuningParameters);
+    rangeBounds = asdk_CGRectExpandHorizontally(rangeBounds,
+                                                horizontalBuffer.negativeDirection,
+                                                horizontalBuffer.positiveDirection);
+    // Update bounds is at most 95% of the next/previous screenful and at least half of tuning parameter value.
+    updateBounds = asdk_CGRectExpandHorizontally(updateBounds,
+                                                 MIN(horizontalBuffer.negativeDirection * 0.5, 0.95),
+                                                 MIN(horizontalBuffer.positiveDirection * 0.5, 0.95));
+  }
+  
+  BOOL canScrollVertically = ASScrollDirectionContainsVerticalDirection(scrollableDirections);
+  if (canScrollVertically) {
+    ASDirectionalScreenfulBuffer verticalBuffer = ASDirectionalScreenfulBufferVertical(scrollDirection,
+                                                                                       rangeTuningParameters);
+    rangeBounds = asdk_CGRectExpandVertically(rangeBounds,
+                                              verticalBuffer.negativeDirection,
+                                              verticalBuffer.positiveDirection);
+    // Update bounds is at most 95% of the next/previous screenful and at least half of tuning parameter value.
+    updateBounds = asdk_CGRectExpandVertically(updateBounds,
+                                               MIN(verticalBuffer.negativeDirection * 0.5, 0.95),
+                                               MIN(verticalBuffer.positiveDirection * 0.5, 0.95));
+  }
+
+  return {rangeBounds, updateBounds};
+}
+
+- (NSSet *)indexPathsForItemsWithinRangeBounds:(CGRect)rangeBounds collectionView:(ASCollectionView *)collectionView {
+  NSMutableSet *indexPathSet = [[NSMutableSet alloc] init];
+  NSArray *layoutAttributes = [collectionView.collectionViewLayout layoutAttributesForElementsInRect:rangeBounds];
+  for (UICollectionViewLayoutAttributes *la in layoutAttributes) {
```

Comment:
Dropping note here for future reference. The iteration is necessary because the item's index path is stored within the layout attributes object. 

In the future, we could define a method that layouts can optionally implement that returns just the index paths to avoid the extra iteration.